Variable bottom topography for the covariant form (and new test cases)

tristanmontoya opened the pull request:
This PR adds the variable bottom topography term $ghG^{ij}\partial_j b$ to the pressure gradient in the non-symmetric two-point flux, where the bottom topography is stored as an auxiliary node variable. The following test cases are included (the last one does not use bottom topography):

- Unsteady solid body rotation (`initial_condition_solid_body_rotation`) from [Läuter et al. (2005)](https://doi.org/10.1016/j.jcp.2005.04.022)
- Zonal flow over an isolated mountain (`initial_condition_isolated_mountain`) from [Williamson et al. (1992)](https://doi.org/10.1016/S0021-9991(05)80016-6)
- Well-balancedness check for isolated mountain topography (same as above but with zero initial velocity)
- Barotropic instability (`initial_condition_barotropic_instability`) from [Galewsky et al. (2004)](https://www.tandfonline.com/doi/abs/10.3402/tellusa.v56i5.14436)

amrueda commented on src/equations/equations.jl:
```diff
@@ -160,17 +160,20 @@ end
            nvars_area_element +
            nvars_metric_covariant +
            nvars_metric_contravariant +
+           nvars_bottom_topography +
```
Is there a reason why you add this to all equations in covariant form? `CovariantLinearAdvectionEquation2D` does not need this variable, and a future implementation of `CovariantCompressibleEulerEquation2D` won't need it either. Is there an easy way to extend `AbstractCovariantEquations` and store the additional quantities only for the equations that need them?

amrueda replied:
I mean, if this is an easy change, it would be nice to do it here. If it's difficult, we can think about the best way to do it and postpone it to a future PR.

tristanmontoya replied:
What I plan to do in the future is replace the bottom topography auxiliary variable with more general "auxiliary fields" that could also include the base states for the perturbation form, and each equation set would have a certain number of auxiliary fields. But for now, I propose leaving that for later, in accordance with the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

amrueda commented on src/equations/covariant_shallow_water_split.jl:
```diff
@@ -112,21 +111,45 @@ end
+                   J_ll * (geometric_term[2] + pressure_term[2]))
 end
 
+# For smooth bottom topography, we can significantly simplify the nonconservative surface
+# term, such that only the pressure term remains.
+@inline function flux_nonconservative_surface_simplified(u_ll, u_rr, aux_vars_ll,
+                                                         aux_vars_rr,
+                                                         orientation::Integer,
+                                                         equations::SplitCovariantShallowWaterEquations2D)
+    # Geometric variables
+    Gcon_ll = metric_contravariant(aux_vars_ll, equations)
+    J_ll = area_element(aux_vars_ll, equations)
+
+    # Physical variables
+    h_ll = waterheight(u_ll, equations)
+    h_rr = waterheight(u_rr, equations)
+
+    pressure_term = equations.gravity * Gcon_ll[:, orientation] * h_ll * h_rr
+
+    return SVector(zero(eltype(u_ll)), J_ll * pressure_term[1], J_ll * pressure_term[2])
+end
```
I understand that this is equivalent to `flux_nonconservative_ec` for smooth bottom topographies (continuous across element interfaces). I suggest to state this more clearly in the comments (maybe a docstring would be useful) and the elixirs, and to make a copy of one of the tests (for example [this one](https://github.com/trixi-framework/TrixiAtmo.jl/pull/73/files#diff-87d10937ca36bcc459b4cd08fb7ae058e3c2ee55cc83e76152083d2c1b49b29cR40-R56)), but using `flux_nonconservative_ec` in the surface.

amrueda commented on src/equations/covariant_shallow_water_split.jl:
```diff
@@ -103,7 +102,7 @@ end
 end
 
-# Non-symmetric part of entropy-conservative flux
+# Non-symmetric part of entropy-conservative flux. Can be used in both surface and volume.
```
Maybe add that the treatment of the bottom topography is consistent with the non-conservative flux of Wintermeyer for flat manifolds?

amrueda reviewed (commented):
Great work, @tristanmontoya! 🙂 
I have some comments:

tristanmontoya commented:
@amrueda I think I've addressed all your comments now. Mostly, this involved improving/adding docstrings to adequately explain what I did and why. I also added a well-balancedness check elixir, which uses the full (non-simplified) `flux_nonconservative_ec` for the surface terms, so now both versions are tested.

amrueda commented on examples/elixir_shallowwater_covariant_well_balanced.jl:
```diff
@@ -0,0 +1,87 @@
+save_solution = SaveSolutionCallback(dt = (tspan[2] - tspan[1]) / n_saves,
+                                     solution_variables = cons2prim_and_vorticity)
+
+# The StepsizeCallback handles the re-calculation of the maximum Δt after each time step.
+# Here we take a small CFL value to make the time integration error negligible when
+# verifying well balancedness.
+stepsize_callback = StepsizeCallback(cfl = 0.01)
```
I don't quite understand this. Why should the time integration affect well-balancedness? If `rhs = 0` for a lake-at-rest solution, then the time-integration method shouldn't affect the constant state... Or am I missing something?

tristanmontoya replied:
Your argument makes sense to me; I will try looking what the max value of `du` is at each time step to see what is going on here.

tristanmontoya replied:
Substituting the initial well balanced state gives a time derivative on the order of machine precision, so I would conclude that the scheme is well balanced. It seems like there are some roundoff error dynamics going on that result in more amplification of the floating-point error for large time steps.

amrueda reviewed (dismissed):
LGTM! Thanks, @tristanmontoya!

Just one minor comment:
